Can't build a RNW project with Visual Studio 2022 17.1

### Problem Description

Since I upgraded my Visual Studio 2022 installation to 17.1, I can't compile anymore any React Native for Windows project. I always get the following error during the build process of the Microsoft.ReactNative project:

C:\src\VS2022\node_modules\react-native-windows\Mso\dispatchQueue\dispatchQueue.h(728,17): error C2607: static assertion failed
C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(486,5): error MSB4181: The "CompileXaml" task returned false but did not log an error.

The same project compiled with Visual Studio 2019 builds fine.

### Steps To Reproduce

 1. Create a new RNW project based on RN 0.67
 2. Open the solution inside the windows folder with Visual Studio 2022 17.1
 3. Build the Microsoft.ReactNative project

### Expected Results

The solution can be compiled without errors

### CLI version

6.4.0

### Environment

```markdown
System:
    OS: Windows 10 10.0.22000
    CPU: (8) x64 Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
    Memory: 2.09 GB / 15.60 GB
  Binaries:
    Node: 14.18.3 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.15 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.15 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.17134.0, 10.0.17763.0, 10.0.18362.0, 10.0.19041.0, 10.0.22000.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.11.32126.315 (Visual Studio Enterprise 2019), 17.2.32210.308 (Visual Studio Enterprise 2022), 17.1.32210.238 (Visual Studio Enterprise 2022)
  Languages:
    Java: 11.0.14 - C:\Program Files\Microsoft\jdk-11.0.14.9-hotspot\bin\javac.EXE
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2
    react-native: 0.67.2 => 0.67.2
    react-native-windows: 0.67.1 => 0.67.1
  npmGlobalPackages:
    *react-native*: Not Found
```


### Target Platform Version

10.0.19041

### Target Device(s)

Desktop

### Visual Studio Version

Visual Studio 2022

### Build Configuration

Debug

### Snack, code example, screenshot, or link to a repository

_No response_

--- a/change/react-native-windows-5e2cc7a5-b37d-49cc-b01b-c3669a0686c3.json
+++ b/change/react-native-windows-5e2cc7a5-b37d-49cc-b01b-c3669a0686c3.json
@@ -1,0 +1,7 @@
+{
+  "type": "prerelease",
+  "comment": "MustBeNoExceptVoidFunctor should depend on the template type parameter to avoid being evaluated too early",
+  "packageName": "react-native-windows",
+  "email": "asklar@microsoft.com",
+  "dependentChangeType": "patch"
+}

--- a/vnext/Mso/dispatchQueue/dispatchQueue.h
+++ b/vnext/Mso/dispatchQueue/dispatchQueue.h
@@ -723,9 +723,14 @@ inline DispatchTaskImpl<TInvoke, TOnCancel>::~DispatchTaskImpl() noexcept {
   }
 }
 
+namespace details {
+template <typename>
+constexpr bool always_false = false;
+}
+
 template <typename T>
 inline void MustBeNoExceptVoidFunctor() {
-  static_assert(false, __FUNCTION__ ": not a noexcept callable functor returning void");
+  static_assert(details::always_false<T>, __FUNCTION__ ": not a noexcept callable functor returning void");
 }
 
 template <typename TInvoke, typename TOnCancel>

